Terminal.Gui/Core/Application.cs

```diff
@@ -279,9 +279,22 @@ public override void Post (SendOrPostCallback d, object state)
 
 			public override void Send (SendOrPostCallback d, object state)
 			{
-				mainLoop.Invoke (() => {
-					d (state);
-				});
+				if (Thread.CurrentThread.ManagedThreadId == _mainThreadId) {
+					mainLoop.Invoke (() => {
+						d (state);
+					});
+				} 
+				else 
+				{
+					var wasExecuted = false;
+					mainLoop.Invoke (() => {
+						d (state);
+						wasExecuted = true;
+					});
+					while (!wasExecuted) {
+						Thread.Sleep (15);
+					}
```

Comment:
@BDisp: The main UI thread is not blocked by this while loop. In case the `Send `method is called from the UI thread the given `action` parameter is invoked immediatly. In case the `Send `method is called from any other thread instead of the UI main thread the while loop will be executed until the idle handler of the mainloop invoked the given `action` successfully.